### PR TITLE
ICU-21349 Remove a potential 0/0 (=NaN). Align C++ & Java better.

### PR DESCRIPTION
**Pull request by hugovdm**
This improves some unit tests and makes C++ and Java code more similar, with an edge-case improvement in C++ complex number conversion accuracy.

I was considering tackling icu-units#108, but I think we should discuss what precision-handling differences we want between C++ and Java.

##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21349
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [X] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/i18n/units_complexconverter.cpp [hunk 1]
```diff
@@ -132,24 +132,20 @@ MaybeStackVector<Measure> ComplexUnitsConverter::convert(double quantity,
...
-            // whether to use the current unit or the next smaller unit, we
-            // therefore nudge up the number with which the thresholding
-            // decision is made. However after the thresholding, we use the
-            // original values to ensure unbiased accuracy (to the extent of
-            // double's capabilities).
-            int64_t roundedQuantity = floor(quantity * (1 + DBL_EPSILON));
-            intValues[i] = roundedQuantity;
+            // If quantity is at the limits of double's precision from an
+            // integer value, we take that integer value.
+            int64_t flooredQuantity = floor(quantity * (1 + DBL_EPSILON));
```
Comment (hugovdm) [inline icu4c/source/i18n/units_complexconverter.cpp:137]: Now that we have support for a `RoundingImpl *rounder` parameter, we could probably drop this DBL_EPSILON nudge by setting a default `rounder` instead. I'm happy to add it to this PR if you prefer, otherwise I can do this in a later PR.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/units/ComplexUnitsConverter.java [hunk 2]
```diff
@@ -129,15 +129,17 @@ public boolean greaterThanOrEqual(BigDecimal quantity, BigDecimal limit) {
...
+                BigDecimal flooredQuantity =
                     quantity.multiply(EPSILON_MULTIPLIER).setScale(0, RoundingMode.FLOOR);
-                intValues.add(roundedQuantity);
+                intValues.add(flooredQuantity);
 
                 // Keep the residual of the quantity.
                 //   For example: `3.6 feet`, keep only `0.6 feet`
-                quantity = quantity.subtract(roundedQuantity);
-                if (quantity.compareTo(BigDecimal.ZERO) == -1) {
+                BigDecimal remainder = quantity.subtract(flooredQuantity);
```
Comment (hugovdm) [inline icu4j/main/classes/core/src/com/ibm/icu/impl/units/ComplexUnitsConverter.java:138]: Looking at this again, I do notice I'm adding a new variable without much reason - I'd be happy to change this back to just using quantity directly if someone prefers?
(`quantity = quantity.subtract(flooredQuantity);` for J, and `quantity -= flooredQuantity;` for C.)

**Review by sffc — APPROVED** [review]
